graph/graph.go

```diff
@@ -402,6 +404,43 @@ func (graph *Graph) ByParent() (map[string][]*image.Image, error) {
 	return byParent, err
 }
 
+// If the images and layers are in pulling chain, mark them.
+// If not, they may be deleted by rmi with dangling condition.
+func (graph *Graph) AddtoPulling(layerIDs []string, imgID string) {
+	for _, layerID := range layerIDs {
+		if imgs, exists := graph.pullingImgs[layerID]; exists {
+			graph.pullingImgs[layerID] = append(imgs, imgID)
+		} else {
+			graph.pullingImgs[layerID] = []string{imgID}
+		}
+	}
+}
+
+// If the images and layers have benn pulled, unmark them.
+func (graph *Graph) DelfromPulling(layerIDs []string, imgID string) {
+	for _, layerID := range layerIDs {
+		if imgs, exists := graph.pullingImgs[layerID]; exists {
+			result := make([]string, 0)
+			index := 0
+			endIndex := len(imgs) - 1
+			for k, v := range imgs {
+				if v == imgID {
+					result = append(result, imgs[index:k]...)
+					index = k + 1
+				} else if k == endIndex {
+					result = append(result, imgs[index:endIndex+1]...)
+				}
+			}
+			if current := len(result); current == 0 {
+				delete(graph.pullingImgs, layerID)
+			} else {
+				imgs = make([]string, len(result))
```

Comment:
if length of "graph.pullingImgs[layerID]" is bigger than length fo result, copy(graph.pullingImgs[layerID], result) will make mistake.
But the original code actually did not update the map.
I made a little change.
Please review.
Thanks!